src/header/common/strict_transport_security.rs

```diff
@@ -0,0 +1,259 @@
+        self.parse_tokens()
+    }
+
+    fn parse_quoted_string(&mut self) -> String {
+        let mut result = String::new();
+        self.pop();
+
+        loop {
+            let next = match self.pop() {
+                '\\' => self.pop(),
+                '"' => break,
+                c @ _ => c
+            };
+
+            result.push(next)
+        }
+
+        self.pop_whitespace();
+
+        result
+    }
+
+    fn parse_directive_value(&mut self) -> Option<String> {
+        self.pop_whitespace();
+        if self.eof() { return None }
+
+        match self.peek() {
+            '=' => {
+                self.pop();
+                self.pop_whitespace();
+
+                if self.eof() { return None }
+
+                let result = match self.peek() {
+                    '"' => Some(self.parse_quoted_string()),
+                    _ => Some(self.parse_value())
+                };
+
+                self.pop_to_sep();
+
+                result
+            }
+            _ => {
+                self.pop_to_sep();
+
+                None
+            }
+        }
+    }
+
+    fn parse_directive(&mut self) -> Option<StsDirective> {
+        let directive_name = self.parse_directive_name().to_lowercase();
+
+        match &*directive_name {
+            "max-age" => {
+                match self.parse_directive_value() {
+                    Some(max_age_val) => {
+                        match max_age_val.parse::<u64>() {
+                            Ok(max_age) => Some(StsDirective::MaxAge(max_age)),
+                            _ => None
+                        }
+                    },
+                    None => None
+                }
+            }
+            "includesubdomains" => Some(StsDirective::IncludeSubdomains(true)),
+            _ => None
+        }
+    }
+
+    fn parse(input: Vec<u8>) -> ::Result<StrictTransportSecurity> {
+        let mut parser = StsParser { pos: 0, input: input };
+        let mut directives = Vec::new();
+
+        while !parser.eof() {
+            directives.push(parser.parse_directive());
+        }
+
+        let (include_subdomains, max_age) = directives.iter().fold((None, None), |m, d| match d {
+            &Some(StsDirective::MaxAge(a)) => (m.0, Some(a)),
+            &Some(StsDirective::IncludeSubdomains(i)) => (Some(i), m.1),
+            _ => m
+        });
+
+        match (include_subdomains, max_age) {
+            (Some(_), Some(max_age)) => Ok(StrictTransportSecurity(true, max_age)),
+            (None, Some(max_age)) => Ok(StrictTransportSecurity(false, max_age)),
+            _ => Err(::Error::Header)
+        }
+    }
+}
+
+impl Header for StrictTransportSecurity {
+    fn header_name() -> &'static str {
+        "Strict-Transport-Security"
+    }
+
+    fn parse_header(raw: &[Vec<u8>]) -> ::Result<StrictTransportSecurity> {
+        if let Some(first_header_raw) = raw.iter().nth(0) {
+            StsParser::parse(first_header_raw.clone())
```

Comment:
Could the parser do the parsing with just a slice, instead of cloning the `Vec`?